pkg/controller/core/clusterqueue_controller.go

```diff
@@ -291,18 +295,47 @@ func (r *ClusterQueueReconciler) SetupWithManager(mgr ctrl.Manager) error {
 		Complete(r)
 }
 
-func (r *ClusterQueueReconciler) Status(cq *kueue.ClusterQueue) (kueue.ClusterQueueStatus, error) {
+func (r *ClusterQueueReconciler) UpdateCQStatusIfChanged(
+	ctx context.Context,
+	cq *kueue.ClusterQueue,
+	conditionStatus metav1.ConditionStatus,
+	reason, msg string,
+) error {
 	usage, workloads, err := r.cache.Usage(cq)
 	if err != nil {
 		r.log.Error(err, "Failed getting usage from cache")
 		// This is likely because the cluster queue was recently removed,
 		// but we didn't process that event yet.
-		return kueue.ClusterQueueStatus{}, err
+		return err
 	}
+	pendingWorkloads := r.qManager.Pending(cq)
+	oldStatus := cq.Status.DeepCopy()
+	cq.Status.UsedResources = usage
+	cq.Status.AdmittedWorkloads = int32(workloads)
+	cq.Status.PendingWorkloads = int32(pendingWorkloads)
+
+	activeConditionIsChanged := UpdateCQConditionsIfChanged(&cq.Status.Conditions, metav1.Condition{
+		Type:               kueue.ClusterQueueActive,
+		Status:             conditionStatus,
+		LastTransitionTime: metav1.Now(),
+		Reason:             reason,
+		Message:            msg,
+	})
+	if activeConditionIsChanged ||
+		!equality.Semantic.DeepEqual(cq.Status.UsedResources, oldStatus.UsedResources) ||
+		cq.Status.AdmittedWorkloads != oldStatus.AdmittedWorkloads ||
+		cq.Status.PendingWorkloads != oldStatus.PendingWorkloads {
+		return r.client.Status().Update(ctx, cq)
+	}
+	return nil
+}
 
-	return kueue.ClusterQueueStatus{
-		UsedResources:     usage,
-		AdmittedWorkloads: int32(workloads),
-		PendingWorkloads:  int32(r.qManager.Pending(cq)),
-	}, nil
+func UpdateCQConditionsIfChanged(
+	cqConditions *[]metav1.Condition,
+	condition metav1.Condition,
+) bool {
+	oldCondition := meta.FindStatusCondition(*cqConditions, kueue.ClusterQueueActive).DeepCopy()
+	meta.SetStatusCondition(cqConditions, condition)
+	newCondition := meta.FindStatusCondition(*cqConditions, kueue.ClusterQueueActive).DeepCopy()
+	return !equality.Semantic.DeepEqual(newCondition, oldCondition)
```

Comment:
I think this is not necessary, given the comment about having a single DeepEqual

Reply:
Makes sense.

Reply:
Done.